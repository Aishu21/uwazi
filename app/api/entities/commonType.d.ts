/* eslint-disable */
/**AUTO-GENERATED. RUN yarn build_schema to update.*/

export type ObjectIdSchema =
  | string
  | {
      [k: string]: any | undefined;
    };
