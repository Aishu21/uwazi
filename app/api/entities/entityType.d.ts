/* eslint-disable */
/**AUTO-GENERATED. RUN yarn emitTypes to update.*/

import {
  ObjectIdSchema,
  LinkSchema,
  DateRangeSchema,
  GeolocationSchema,
  TocSchema,
} from 'shared/commonTypes';

export interface EntitySchema {
  _id?: ObjectIdSchema;
  sharedId?: string;
  language?: string;
  mongoLanguage?: string;
  title?: string;
  template?: ObjectIdSchema;
  file?: {
    originalname?: string;
    filename?: string;
    mimetype?: string;
    size?: number;
    timestamp?: number;
    language?: string;
  };
  fullText?: {
    /**
     * This interface was referenced by `undefined`'s JSON-Schema definition
     * via the `patternProperty` "^[0-9]+$".
     */
    [k: string]: string;
  };
  totalPages?: number;
  icon?: {
    _id?: string;
    label?: string;
    type?: string;
  };
  attachments?: {
    originalname?: string;
    filename?: string;
    mimetype?: string;
    timestamp?: number;
    size?: number;
    [k: string]: any | undefined;
  }[];
  creationDate?: number;
  processed?: boolean;
  uploaded?: boolean;
  published?: boolean;
  pdfInfo?: {
    /**
     * This interface was referenced by `undefined`'s JSON-Schema definition
     * via the `patternProperty` "^[0-9]+$".
     */
    [k: string]: {
      chars?: number;
    };
  };
  toc?: TocSchema[];
  user?: ObjectIdSchema;
  metadata?: {
    [k: string]:
      | (
          | null
          | string
          | number
          | string[]
          | number[]
          | (number | null)[]
          | DateRangeSchema
          | DateRangeSchema[]
          | LinkSchema
          | GeolocationSchema)
      | undefined;
  };
  [k: string]: any | undefined;
}
