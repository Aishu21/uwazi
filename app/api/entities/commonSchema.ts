/** @format */

export const objectIdSchema = {
  oneOf: [{ type: 'string' }, { type: 'object' }],
};
