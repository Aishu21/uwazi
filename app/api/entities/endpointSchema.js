import Joi from 'joi';

const schema = Joi.object().keys({
  _id: Joi.string(),
  __v: Joi.number(),
  language: Joi.string(),
  mongoLanguage: Joi.string(),
  sharedId: Joi.string(),
  title: Joi.string(),
  type: Joi.string(),
  template: Joi.string(),
  file: Joi.object().keys({
    originalname: Joi.string(),
    filename: Joi.string(),
    mimetype: Joi.string(),
    size: Joi.number(),
    language: Joi.string()
  }),
  fullText: Joi.any(),
  totalPages: Joi.number(),
  icon: Joi.object().keys({
    _id: Joi.string(),
    label: Joi.string(),
    type: Joi.string()
  }),
  toc: Joi.array().items(Joi.object().keys({
    label: Joi.string(),
    indentation: Joi.number(),
    range: Joi.object().keys({
      start: Joi.number(),
      end: Joi.number()
    })
  })),
  attachments: Joi.array().items(Joi.object().keys({
    originalname: Joi.string(),
    filename: Joi.string(),
    mimetype: Joi.string(),
    size: Joi.number()
  })),
  creationDate: Joi.number(),
  processed: Joi.boolean(),
  uploaded: Joi.boolean(),
  published: Joi.boolean(),
  metadata: Joi.object().keys().pattern(Joi.string(), Joi.alternatives().try(
    Joi.number(),
    Joi.string(),
    Joi.array().items(Joi.alternatives().try(
      Joi.number(),
      Joi.string()
    )),
    Joi.array().items(Joi.object().pattern(Joi.string(),
      Joi.array().items(Joi.string())))
  )),
  pdfInfo: Joi.any(),
  user: Joi.string()
}).required();

export default schema;
