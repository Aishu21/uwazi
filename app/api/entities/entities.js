import {db_url as dbURL} from 'api/config/database';
import request from 'shared/JSONRequest';
import {updateMetadataNames, deleteMetadataProperties} from 'api/entities/utils';
import date from 'api/utils/date.js';
import sanitizeResponse from '../utils/sanitizeResponse';
import references from '../references/references.js';
import settings from '../settings';
import ID from 'shared/uniqueId';

export default {
  save(doc, {user, language}) {
    doc.type = 'entity';
    if (!doc._id) {
      doc.user = user;
      doc.creationDate = date.currentUTC();
    }

    const sharedId = doc.sharedId || ID();
    return settings.get()
    .then(({languages}) => {
      if (doc._id) {
        return request.post(dbURL + '/_design/entities/_update/partialUpdate/' + doc._id, doc);
      }

      const docs = languages.map((lang) => {
        let langDoc = Object.assign({}, doc);
        langDoc.language = lang.key;
        langDoc.sharedId = sharedId;
        return langDoc;
      });

      return request.post(`${dbURL}/_bulk_docs`, {docs});
    })
    .then(() => request.get(`${dbURL}/_design/entities/_view/by_language`, {key: [sharedId, language]}))
    .then(response => {
      return Promise.all([response, references.saveEntityBasedReferences(response.json.rows[0].value)]);
    })
    .then(([response]) => response.json.rows[0].value);
  },

  countByTemplate(templateId) {
    return request.get(`${dbURL}/_design/entities/_view/count_by_template?group_level=1&key="${templateId}"`)
    .then((response) => {
      if (!response.json.rows.length) {
        return 0;
      }
      return response.json.rows[0].value;
    });
  },

  getByTemplate(templateId) {
    return request.get(`${dbURL}/_design/entities/_view/by_template?&key="${templateId}"`)
    .then((response) => {
      return sanitizeResponse(response.json);
    });
  },

  updateMetadataProperties(templateId, nameMatches, deleteProperties) {
    return request.get(`${dbURL}/_design/entities/_view/metadata_by_template?key="${templateId}"`)
    .then((response) => {
      let entities = response.json.rows.map((r) => r.value);
      entities = updateMetadataNames(entities, nameMatches);
      entities = deleteMetadataProperties(entities, deleteProperties);

      let updates = [];
      entities.forEach((entity) => {
        let url = `${dbURL}/_design/entities/_update/partialUpdate/${entity._id}`;
        updates.push(request.post(url, entity));
      });

      return Promise.all(updates);
    });
  },

  delete(id) {
    let docsToDelete = [];
    return request.get(`${dbURL}/${id}`)
    .then((response) => {
      docsToDelete.push({_id: response.json._id, _rev: response.json._rev});
      return request.get(`${dbURL}/_design/references/_view/by_source?key="${id}"`);
    })
    .then((response) => {
      sanitizeResponse(response.json);
      docsToDelete = docsToDelete.concat(response.json.rows);
      return request.get(`${dbURL}/_design/references/_view/by_target?key="${id}"`);
    })
    .then((response) => {
      sanitizeResponse(response.json);
      docsToDelete = docsToDelete.concat(response.json.rows);
      docsToDelete.map((doc) => doc._deleted = true);
      return request.post(`${dbURL}/_bulk_docs`, {docs: docsToDelete});
    })
    .then((response) => {
      return response.json;
    });
  }
};
