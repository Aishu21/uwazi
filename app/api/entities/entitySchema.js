/** @format */

import Ajv from 'ajv';
import templatesModel from 'api/templates/templatesModel';
import { templateTypes } from 'shared/templateTypes';
import { isNumber, isUndefined, isString, isObject, isNull } from 'util';
import { objectIdSchema } from 'api/entities/commonSchema';

const ajv = Ajv({ allErrors: true });

const isEmpty = value => isNull(value) || isUndefined(value) || !value.length;

const isNonArrayObject = value => isObject(value) && !Array.isArray(value);

const isValidDateRange = value => {
  if (!isNonArrayObject(value)) {
    return false;
  }
  if (isNumber(value.from) && isNumber(value.to)) {
    return value.from <= value.to;
  }
  return true;
};

const isValidSelect = value => isString(value) && value;

const isValidGeolocation = value =>
  isString(value.label) && isNumber(value.lat) && isNumber(value.lon);

const validateRequiredProperty = (property, value) => {
  if (property.required) {
    return !isEmpty(value);
  }
  return true;
};

const validateTextProperty = (property, value) => {
  if (
    [
      templateTypes.text,
      templateTypes.markdown,
      templateTypes.media,
      templateTypes.image,
      templateTypes.select,
    ].includes(property.type)
  ) {
    return isString(value);
  }
  return true;
};

const validateNumericProperty = (property, value) => {
  if (property.type === templateTypes.numeric) {
    return isNumber(value) || value === '';
  }
  return true;
};

const validateDateProperty = (property, value) => {
  if (property.type === templateTypes.date) {
    return isNumber(value) && value >= 0;
  }
  return true;
};

const validateMultiDateProperty = (property, value) => {
  if (property.type === templateTypes.multidate) {
    return Array.isArray(value) && value.every(isNumber);
  }
  return true;
};

const validateDateRangeProperty = (property, value) => {
  if (property.type === templateTypes.daterange) {
    return isValidDateRange(value);
  }
  return true;
};

const validateMultiDateRangeProperty = (property, value) => {
  if (property.type === templateTypes.multidaterange) {
    return value.every(isValidDateRange);
  }
  return true;
};

const validateGeolocationProperty = (property, value) => {
  if (property.type === templateTypes.geolocation) {
    return Array.isArray(value) && value.every(isValidGeolocation);
  }
  return true;
};

const validateMultiSelectProperty = (property, value) => {
  if ([templateTypes.multiselect, templateTypes.relationship].includes(property.type)) {
    return Array.isArray(value) && value.every(isValidSelect);
  }
  return true;
};

const validateLinkProperty = (property, value) => {
  if (property.type === templateTypes.link) {
    return isString(value.label) && value.label && isString(value.url) && value.url;
  }
  return true;
};

const validateMetadataField = (property, entity) => {
  const value = entity.metadata && entity.metadata[property.name];
  if (!validateRequiredProperty(property, value)) {
    return false;
  }
  if (isUndefined(value) || isNull(value)) {
    return true;
  }
  const propertyValidators = [
    validateDateProperty,
    validateMultiDateProperty,
    validateDateRangeProperty,
    validateMultiDateRangeProperty,
    validateTextProperty,
    validateNumericProperty,
    validateMultiSelectProperty,
    validateLinkProperty,
    validateGeolocationProperty,
  ];
  return propertyValidators.every(validate => validate(property, value));
};

ajv.addKeyword('metadataMatchesTemplateProperties', {
  async: true,
  errors: false,
  type: 'object',
  async validate(schema, entity) {
    if (!entity.template) {
      return true;
    }
    const [template] = await templatesModel.get({ _id: entity.template });
    if (!template) {
      return false;
    }

    return template.properties.every(property => validateMetadataField(property, entity));
  },
});

export const linkSchema = {
  type: 'object',
  required: ['label', 'url'],
  additionalProperties: false,
  properties: {
    label: { type: 'string', minLength: 1 },
    url: { type: 'string', minLength: 1 },
  },
};

export const dateRangeSchema = {
  type: 'object',
  additionalProperties: false,
  properties: {
    from: {
      oneOf: [{ type: 'number' }, { type: 'null' }],
    },
    to: {
      oneOf: [{ type: 'number' }, { type: 'null' }],
    },
  },
};

export const latLonSchema = {
  type: 'object',
  required: ['lon', 'lat'],
  additionalProperties: false,
  properties: {
    label: { type: 'string' },
    lat: { type: 'number', minimum: -90, maximum: 90 },
    lon: { type: 'number', minimum: -180, maximum: 180 },
  },
};

export const geolocationSchema = {
  type: 'array',
  items: latLonSchema,
};

export const tocSchema = {
  type: 'object',
  title: 'TableOfContents',
  additionalProperties: false,
  properties: {
    range: {
      type: 'object',
      additionalProperties: false,
      properties: {
        start: { type: 'number' },
        end: { type: 'number' },
      },
    },
    label: { type: 'string' },
    indentation: { type: 'number' },
  },
};

export const entitySchema = {
  $schema: 'http://json-schema.org/schema#',
  $async: true,
  type: 'object',
  metadataMatchesTemplateProperties: true,
  additionalProperties: false,
  definitions: {
    objectIdSchema: objectIdSchema,
    linkSchema,
    dateRangeSchema,
    latLonSchema,
    geolocationSchema,
    tocSchema,
  },
  properties: {
    _id: objectIdSchema,
    sharedId: { type: 'string', minLength: 1 },
    language: { type: 'string', minLength: 1 },
    mongoLanguage: { type: 'string' },
    title: { type: 'string', minLength: 1 },
    template: objectIdSchema,
    file: {
      type: 'object',
      additionalProperties: false,
      properties: {
        originalname: { type: 'string' },
        filename: { type: 'string' },
        mimetype: { type: 'string' },
        size: { type: 'number' },
        timestamp: { type: 'number' },
        language: { type: 'string' },
      },
    },
    fullText: {
      type: 'object',
      additionalProperties: false,
      patternProperties: {
        '^[0-9]+$': { type: 'string' },
      },
    },
    totalPages: { type: 'number' },
    icon: {
      type: 'object',
      additionalProperties: false,
      properties: {
        _id: { type: 'string' },
        label: { type: 'string' },
        type: { type: 'string' },
      },
    },
    attachments: {
      type: 'array',
      items: {
        type: 'object',
        additionalProperties: false,
        properties: {
          originalname: { type: 'string' },
          filename: { type: 'string' },
          mimetype: { type: 'string' },
          timestamp: { type: 'number' },
          size: { type: 'number' },
        },
      },
    },
    creationDate: { type: 'number' },
    processed: { type: 'boolean' },
    uploaded: { type: 'boolean' },
    published: { type: 'boolean' },
    pdfInfo: {
      type: 'object',
      additionalProperties: false,
      patternProperties: {
        '^[0-9]+$': {
          type: 'object',
          additionalProperties: false,
          properties: {
            chars: { type: 'number' },
          },
        },
      },
    },
    toc: {
      type: 'array',
      items: tocSchema,
    },
    user: objectIdSchema,
    metadata: {
      type: 'object',
      additionalProperties: {
        anyOf: [
          { type: 'null' },
          { type: 'string' },
          { type: 'number' },
          {
            type: 'array',
            items: {
              type: 'string',
            },
          },
          {
            type: 'array',
            items: {
              type: 'number',
            },
          },
          dateRangeSchema,
          {
            type: 'array',
            items: dateRangeSchema,
          },
          linkSchema,
          geolocationSchema,
        ],
      },
    },
  },
};

export const validateEntity = ajv.compile(entitySchema);
