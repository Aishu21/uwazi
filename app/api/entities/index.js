import entities from './entities.js';
import model from './entitiesModel.js';
import endpointSchema from './endpointSchema';
export default entities;

export {
  model,
  endpointSchema
};
