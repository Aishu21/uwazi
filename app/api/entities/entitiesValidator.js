import { validateEntity } from './entitySchema';

const validator = {
  save(entity) {
    return validateEntity(entity);
  }
};

export default validator;
