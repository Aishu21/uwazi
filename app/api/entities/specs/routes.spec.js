import documentRoutes from '../routes.js';
import database from '../../utils/database.js';
import fixtures from './fixtures.js';
import instrumentRoutes from '../../utils/instrumentRoutes';
import entities from '../entities';
import {catchErrors} from 'api/utils/jasmineHelpers';

describe('entities', () => {
  let routes;

  beforeEach((done) => {
    database.reset_testing_database()
    .then(() => database.import(fixtures))
    .then(done)
    .catch(done.fail);
    routes = instrumentRoutes(documentRoutes);
  });

  describe('POST', () => {
    it('should need authorization', () => {
      expect(routes.post('/api/entities')).toNeedAuthorization();
    });

    it('should create a new document with use user', (done) => {
      let req = {
        body: {title: 'Batman begins'},
        user: {_id: 'c08ef2532f0bd008ac5174b45e033c93', username: 'admin'},
        language: 'lang'
      };

      spyOn(entities, 'save').and.returnValue(new Promise((resolve) => resolve('document')));
      routes.post('/api/entities', req)
      .then((document) => {
        expect(document).toBe('document');
        expect(entities.save).toHaveBeenCalledWith(req.body, {user: req.user, language: 'lang'});
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('/api/entities', () => {
    //it('should return a list of entities returned from the list view', (done) => {
      //routes.get('/api/entities')
      //.then((response) => {
        //expect(response.rows.length).toBe(7);
        //expect(response.rows[0].title).toEqual('Batman finishes');
        //expect(response.rows[0]._id).toEqual('8202c463d6158af8065022d9b5014a18');
        //done();
      //})
      //.catch(catchErrors(done));
    //});

    describe('when passing id', () => {
      it('should return matching document', (done) => {
        let req = {query: {_id: '8202c463d6158af8065022d9b5014ccb'}};

        routes.get('/api/entities', req)
        .then((response) => {
          let docs = response.rows;
          expect(docs.length).toBe(1);
          expect(docs[0].title).toBe('Penguin almost done');
          done();
        })
        .catch(catchErrors(done));
      });
    });
  });

  describe('/api/entities/count_by_template', () => {
    it('should return count of entities using a specific template', (done) => {
      spyOn(entities, 'countByTemplate').and.returnValue(new Promise((resolve) => resolve(2)));
      let req = {query: {templateId: 'templateId'}};

      routes.get('/api/entities/count_by_template', req)
      .then((response) => {
        expect(entities.countByTemplate).toHaveBeenCalledWith('templateId');
        expect(response).toEqual(2);
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('DELETE', () => {
    beforeEach(() => {
      spyOn(entities, 'delete').and.returnValue(Promise.resolve({json: 'ok'}));
    });

    it('should use entities to delete it', (done) => {
      let req = {query: {_id: 123, _rev: 456}};
      return routes.delete('/api/entities', req)
      .then(() => {
        expect(entities.delete).toHaveBeenCalledWith(req.query._id);
        done();
      })
      .catch(catchErrors(done));
    });
  });
});
