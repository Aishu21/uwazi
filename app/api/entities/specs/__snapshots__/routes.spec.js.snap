// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities /api/entities/count_by_template should have a validation schema 1`] = `
Object {
  "children": Object {
    "templateId": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`entities DELETE /api/entities should have a validation schema 1`] = `
Object {
  "children": Object {
    "sharedId": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`entities DELETE /api/entities/multiple should have a validation schema 1`] = `
Object {
  "children": Object {
    "sharedIds": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`entities GET should have a validation schema 1`] = `
Object {
  "children": Object {
    "_id": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "omitRelationships": Object {
      "type": "any",
    },
  },
  "type": "object",
}
`;

exports[`entities POST /entities/multipleupdate should have a validation schema 1`] = `
Object {
  "children": Object {
    "ids": Object {
      "flags": Object {
        "sparse": false,
      },
      "items": Array [
        Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      ],
      "type": "array",
    },
    "values": Object {
      "children": Object {
        "icon": Object {
          "children": Object {
            "_id": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "label": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "type": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
          },
          "type": "object",
        },
        "metadata": Object {
          "patterns": Array [
            Object {
              "rule": Object {
                "alternatives": Array [
                  Object {
                    "invalids": Array [
                      Infinity,
                      -Infinity,
                    ],
                    "type": "number",
                  },
                  Object {
                    "invalids": Array [
                      "",
                    ],
                    "type": "string",
                  },
                  Object {
                    "flags": Object {
                      "sparse": false,
                    },
                    "items": Array [
                      Object {
                        "alternatives": Array [
                          Object {
                            "invalids": Array [
                              Infinity,
                              -Infinity,
                            ],
                            "type": "number",
                          },
                          Object {
                            "invalids": Array [
                              "",
                            ],
                            "type": "string",
                          },
                        ],
                        "type": "alternatives",
                      },
                    ],
                    "type": "array",
                  },
                  Object {
                    "flags": Object {
                      "sparse": false,
                    },
                    "items": Array [
                      Object {
                        "patterns": Array [
                          Object {
                            "rule": Object {
                              "flags": Object {
                                "sparse": false,
                              },
                              "items": Array [
                                Object {
                                  "invalids": Array [
                                    "",
                                  ],
                                  "type": "string",
                                },
                              ],
                              "type": "array",
                            },
                            "schema": Object {
                              "invalids": Array [
                                "",
                              ],
                              "type": "string",
                            },
                          },
                        ],
                        "type": "object",
                      },
                    ],
                    "type": "array",
                  },
                ],
                "type": "alternatives",
              },
              "schema": Object {
                "invalids": Array [
                  "",
                ],
                "type": "string",
              },
            },
          ],
          "type": "object",
        },
        "published": Object {
          "falsy": Array [
            false,
          ],
          "flags": Object {
            "insensitive": true,
          },
          "truthy": Array [
            true,
          ],
          "type": "boolean",
        },
        "template": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`entities POST get_raw_page should have a validation schema 1`] = `
Object {
  "children": Object {
    "pageNumber": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "type": "number",
    },
    "sharedId": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`entities POST should have a validation schema 1`] = `
Object {
  "children": Object {
    "__v": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "type": "number",
    },
    "_id": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "attachments": Object {
      "flags": Object {
        "sparse": false,
      },
      "items": Array [
        Object {
          "children": Object {
            "filename": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "mimetype": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "originalname": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "size": Object {
              "invalids": Array [
                Infinity,
                -Infinity,
              ],
              "type": "number",
            },
          },
          "type": "object",
        },
      ],
      "type": "array",
    },
    "creationDate": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "type": "number",
    },
    "file": Object {
      "children": Object {
        "filename": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "language": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "mimetype": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "originalname": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "size": Object {
          "invalids": Array [
            Infinity,
            -Infinity,
          ],
          "type": "number",
        },
      },
      "type": "object",
    },
    "fullText": Object {
      "type": "any",
    },
    "icon": Object {
      "children": Object {
        "_id": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "label": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
        "type": Object {
          "invalids": Array [
            "",
          ],
          "type": "string",
        },
      },
      "type": "object",
    },
    "language": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "metadata": Object {
      "patterns": Array [
        Object {
          "rule": Object {
            "alternatives": Array [
              Object {
                "invalids": Array [
                  Infinity,
                  -Infinity,
                ],
                "type": "number",
              },
              Object {
                "invalids": Array [
                  "",
                ],
                "type": "string",
              },
              Object {
                "flags": Object {
                  "sparse": false,
                },
                "items": Array [
                  Object {
                    "alternatives": Array [
                      Object {
                        "invalids": Array [
                          Infinity,
                          -Infinity,
                        ],
                        "type": "number",
                      },
                      Object {
                        "invalids": Array [
                          "",
                        ],
                        "type": "string",
                      },
                    ],
                    "type": "alternatives",
                  },
                ],
                "type": "array",
              },
              Object {
                "flags": Object {
                  "sparse": false,
                },
                "items": Array [
                  Object {
                    "patterns": Array [
                      Object {
                        "rule": Object {
                          "flags": Object {
                            "sparse": false,
                          },
                          "items": Array [
                            Object {
                              "invalids": Array [
                                "",
                              ],
                              "type": "string",
                            },
                          ],
                          "type": "array",
                        },
                        "schema": Object {
                          "invalids": Array [
                            "",
                          ],
                          "type": "string",
                        },
                      },
                    ],
                    "type": "object",
                  },
                ],
                "type": "array",
              },
            ],
            "type": "alternatives",
          },
          "schema": Object {
            "invalids": Array [
              "",
            ],
            "type": "string",
          },
        },
      ],
      "type": "object",
    },
    "mongoLanguage": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "pdfInfo": Object {
      "type": "any",
    },
    "processed": Object {
      "falsy": Array [
        false,
      ],
      "flags": Object {
        "insensitive": true,
      },
      "truthy": Array [
        true,
      ],
      "type": "boolean",
    },
    "published": Object {
      "falsy": Array [
        false,
      ],
      "flags": Object {
        "insensitive": true,
      },
      "truthy": Array [
        true,
      ],
      "type": "boolean",
    },
    "sharedId": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "template": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "title": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "toc": Object {
      "flags": Object {
        "sparse": false,
      },
      "items": Array [
        Object {
          "children": Object {
            "indentation": Object {
              "invalids": Array [
                Infinity,
                -Infinity,
              ],
              "type": "number",
            },
            "label": Object {
              "invalids": Array [
                "",
              ],
              "type": "string",
            },
            "range": Object {
              "children": Object {
                "end": Object {
                  "invalids": Array [
                    Infinity,
                    -Infinity,
                  ],
                  "type": "number",
                },
                "start": Object {
                  "invalids": Array [
                    Infinity,
                    -Infinity,
                  ],
                  "type": "number",
                },
              },
              "type": "object",
            },
          },
          "type": "object",
        },
      ],
      "type": "array",
    },
    "totalPages": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "type": "number",
    },
    "type": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "uploaded": Object {
      "falsy": Array [
        false,
      ],
      "flags": Object {
        "insensitive": true,
      },
      "truthy": Array [
        true,
      ],
      "type": "boolean",
    },
    "user": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;
